Add a context menu to rename or delete a heap snapshot from the list

kenzieschmoll opened the pull request:
Fixes https://github.com/flutter/devtools/issues/5954. 

![rename-or-delete](https://github.com/flutter/devtools/assets/43759233/e5ad5891-8617-4640-a9e0-e94955985b44)

polina-c commented:
Thanks for putting it together. This way the feature does not look to be discoverable. Why do not we add icon with pen to the selected snapshot? Or, we can add context menu (vertical `...`) with two items: Delete, Rename.

kenzieschmoll commented:
A user may want to edit a snapshot name even if it is not selected, so for this reason, having an icon that only appears for the selected snapshot may also have usability issues. Adding an overflow menu (vertical `...`) for each snapshot item could work, though the UI may look busy with that icon on each row.

I would argue that right click is discoverable - double click was admittedly a bonus item that users may or may not find.
CC @InMatrix

polina-c commented:
Renaming is not needed very often. So, I think it is ok to show the icon just for selected row, the same way as we do it for table:

<img width="527" alt="Screenshot 2023-07-06 at 2 54 18 PM" src="https://github.com/flutter/devtools/assets/12115586/118bb837-a8d8-4fe1-a2a8-24873d7f8a6a">

Right click is discoverable in general, but not in DevTools, where we do not utilize it a lot. Even if user discovers it, they will not expect second row to become editable after selecting 'Rename' for first row.

Also, implementation with  vertical `...` will be conciser in code, and thus cheaper to support.

kenzieschmoll commented:
> they will not expect second row to become editable after selecting 'Rename' for first row.

This is not the case. Only the row that was right clicked will become editable. 

I will look into adding the overflow menu icon for the selected row, but I don't see an issue with also leaving the right click functionality. We haven't used right click much in the past because flutter's context menu support for web wasn't as good as it is now.

polina-c commented:
> I will look into adding the overflow menu icon for the selected row, but I don't see an issue with also leaving the right click functionality. We haven't used right click much in the past because flutter's context menu support for web wasn't as good as it is now. 

Would it be possible to incapsulate right click handling into mixin somehow? So that it does not crowd the code of the widget, where it is used? 

BTW, in modern cross-plat UI, context actions are more and more behind vertical ellipsis.
So, it is becoming user habit to assume there is no actions, if there is no ellipsis:

<img width="851" alt="Screenshot 2023-07-07 at 8 31 44 AM" src="https://github.com/flutter/devtools/assets/12115586/9bd0f6ea-d75d-45f8-8779-b29525911944">

polina-c commented:
> This is not the case. Only the row that was right clicked will become editable.

On the attached video I see click on 'Rename' just one time, and then I see two snapshots are renamed. How it is?

kenzieschmoll commented:
> On the attached video I see click on 'Rename' just one time, and then I see two snapshots are renamed. How it is?

This was from double clicking in the second row, which I've since removed.

kenzieschmoll commented:
Okay this is ready for another review. I removed the right click and double click functionality in favor of the context menu. The benefit of maintaining both affordances is not worth the cost. Updated the PR description.

polina-c commented on packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/snapshot_list.dart:
```diff
@@ -106,52 +107,195 @@ class _ListControlPane extends StatelessWidget {
+
+  final bool selected;
+
+  final ValueListenable<int?> editIndex;
```
Can you add doc comment for the new members. Some are not obvious (editIndex, onEdit vs onEditComplete)

kenzieschmoll replied:
done

polina-c reviewed (approved):
LGTM after comments
